Pool-Based DEX Swap Standard (#112)
onflow/flips: application/20230522-pool-based-dex.md

---
status: proposed 
flip: NNN (do not set)
authors: Satyam Agrawal (satyam.agrawal@dapperlabs.com)
sponsor: Jonny (Increment.Fi) (hsyjonny@gmail.com) 
updated: 2023-06-09 
---

# Pool-Based DEX Swap Standard

## Objective

> The Flow ecosystem is experiencing continued growth and with this expansion comes a need to expand the DeFi ecosystem on the Flow blockchain. Expansion of DeFi on Flow requires new liquidity - typically through cross-chain bridges - as well as straightforward integration for wallets and developers to provide a simple and intuitive user experience. While the macro topic of liquidity is beyond the scope of this discussion, the initial focus of the nascent standard aims to enable ease of interoperability with, and composability of, DEXs which are a cornerstone concept serving as gateways to any DeFi ecosystem. This FLIP is focused on the pool based DEXs like Uniswap style AMMs.

Reviewer: When referring to uniswap, need to indicate which version. V2 is very different from v3 and the proposed v4 adds conveniences to v3.

Author: Thanks for the pointer. It is fixed

## User Benefit

This FLIP primarily benefits developers aiming to integrate DEXs and construct composable financial products within the FLOW blockchain. The proposal's fundamental advantage lies in establishing a unified interface, alleviating developers' concerns regarding the nuances of individual DEX interfaces. This streamlined approach ensures that the integration process becomes more efficient, promoting a plug-and-play model for developers. Consequently, even when new DEXs are introduced into the ecosystem, they will be immediately compatible and supportable, courtesy of the standardized interface. This forward-looking design promises to facilitate continual innovation, bolstering the FLOW ecosystem's growth and dynamism.

## Design Proposal

> The crux of this proposal revolves around the introduction of two distinct resource interfaces. The first, `ImmediateSwap`, is dedicated to facilitating direct trades between two tokens. The second, `ImmediateSwapQuotation`, is designed to retrieve the current price data for a particular token pair. In accordance with this proposal, these two resource interfaces are expected to be adhered to within the resource implementation carried out by various DEXs. This standardization ensures a consistent, interoperable approach to both trading and price quotation across different decentralized exchanges within the ecosystem.

Reviewer: TWAP interface would be very beneficial for future-proofing this Interface. Price Oracle hacks have been devastating in defi. Time waited average price, this was an addition to uniswap to help mitigate oracle attacks with lending protocols. Primarily against flashloan exploits.

Author: Yeah I agree, But don't you think it would be more of an implementation detail whether they are recording the trade value or not for a given block and when queried provide the weighted average of the values? But I do agree we are not providing an interface that offers the API to query that value. However, even with the proposed interface, dex implementers are not restricted to the TWAP-supported implementation of the DEX. 

This proposal's intention is to be very light weighted as much as possible. Maybe once we get enough requirements from the community, we can propose an extension. 

Although I am not completely sure.

Reviewer 2: Agreed that it's important, but I do think it's an implementation detail. There's nuance even in TWAP implementation - periodicity, variability of that periodicity (GARCH), etc. Looks like Increment has an example of a fixed-window TWAP oracle. So maybe we explicitly note that details such as price determinants such as TWAP are up to the implementer and resources exist within the ecosystem for use as examples and is up to the discretion of the developers?

The functions proposed in this document leverage an off-chain computed optimized path array. This methodology is adopted due to the limited likelihood of possessing a direct pool between a given pair of tokens. In order to execute a successful swap, a path from the source token to the target token is required, wherein all adjacent pairs within the path represent potential pool pairs.

Computing such an optimized path on-chain would not be feasible due to the potentially extensive number of variations, making it impractical and inefficient in terms of gas usage. Therefore, it falls upon the transaction initiator to provide a valid path to ensure the successful completion of the transaction. This approach not only optimizes the use of resources but also maintains the robustness and reliability of the transaction process within the ecosystem.

```cadence

/// Data structure that get returned after the `swapExactSourceToTargetTokenUsingPathAndReturn` function execution.
pub struct interface ExactSwapAndReturnValue {
    /// It represents the Vault that holds target token and would be returned
    /// after a swap. 

>     pub let targetTokenVault: @FungibleToken.Vault

Author: Yeah nice catch

    /// It is an optional vault that holds the leftover source tokens after a swap.
    pub let remainingSourceTokenVault: @FungibleToken.Vault?
}

pub resource interface ImmediateSwap {

>     /// @notice It will Swap the source token for to target token

Reviewer 3: ```suggestion
    /// @notice It will Swap the source token for the target token
```

    /// USDC would be the source token
    ///
    /// Necessary constraints
    /// - For the given source vault balance, Swapped target token amount should be
    ///   greater than or equal to `exactTargetAmount`, otherwise swap would fail.
    /// - If the swap settlement time i.e getCurrentBlock().timestamp is less than or
    ///   equal to the provided expiry then the swap would fail.
    /// - Provided `recipient` capability should be valid otherwise the swap would fail.
    /// - If the provided path doesn’t exists then the swap would fail.
    ///
    /// @param sourceToTargetTokenPath: Off-chain computed path for reaching source token to target token
    ///                                 `sourceToTargetTokenPath[0]` should be the source token type while
    ///                                 `sourceToTargetTokenPath[sourceToTargetTokenPath.length - 1]` should be the target token
    ///                                 and all the remaining intermediaries token types would be necessary swap hops to swap the
    ///                                 source token with target token.
    /// @param sourceVault:             Vault that holds the source token.

>     /// @param exactTargetAmount:       Exact amount expected from the swap, If swapped amount is less than `exactTargetAmount` then

Reviewer 3: Is there a way to enforce this in a post condition?

Author: I don't think it can be enforced because we are not returning anything and neither we have the resource variable to compare with.

    ///                                 function execution would throw a error.
    /// @param expiry:                  Unix timestamp after which trade would get invalidated.
    /// @param recipient:               A valid capability that receives target token after the completion of function execution.
    /// @param remainingSourceTokenRecipient: A valid capability that receives surplus source token after the completion of function execution.
	pub fun swapExactSourceToTargetTokenUsingPath(
		sourceToTargetTokenPath: Type[],
		sourceVault: @FungibleToken.Vault,
		exactTargetAmount: UFix64,
		expiry: UFix64,
		recipient: Capability<&{FungibleToken.Receiver}>,

>         remainingSourceTokenRecipient: Capability<&{FungibleToken.Receiver}>

Author: Somehow it is not showing on my system

    /// and wants to perform some actions using the receiving amount.
    ///
    /// Necessary constraints
    /// - For the given source vault balance, Swapped target token amount should be
    ///   greater than or equal to exactTargetAmount, otherwise swap would fail
    /// - If the swap settlement time i.e getCurrentBlock().timestamp is less than or equal to the provided expiry then the swap would fail
    /// - If the provided path doesn’t exists then the swap would fail.
    ///
    /// @param sourceToTargetTokenPath: Off-chain computed path for reaching source token to target token
    ///                                 `sourceToTargetTokenPath[0]` should be the source token type while
    ///                                 `sourceToTargetTokenPath[sourceToTargetTokenPath.length - 1]` should be the target token
    ///                                 and all the remaining intermediaries token types would be necessary swap hops to swap the
    ///                                 source token with target token.
    /// @param sourceVault:             Vault that holds the source token.

>     /// @param exactTargetAmount:       Exact amount expected from the swap, If swapped amount is less than `exactTargetAmount` then

Reviewer 3: We can verify this in a post condition, right?

Author: Same applies here

```cadence
   /// Below event get emitted during `swapExactSourceToTargetTokenUsingPath` & `swapExactSourceToTargetTokenUsingPathAndReturn`
   /// function call.
   ///
   /// @param senderAddress:    Address who initiated the swap. // TODO: It is not possible to know as per the proposed interface.
   /// @param receiverAddress:  Address who receives target token, It is optional because in case of 
   /// `swapExactSourceToTargetTokenUsingPathAndReturn` function target token vault would be returned instead of movement of funds in
   ///  receiver capability.
   /// @param sourceTokenAmount: Amount of sourceToken sender wants to swap.
   /// @param receivedTargetTokenAmount: Amount of targetToken receiver would receive after swapping given `sourceTokenAmount`.
   /// @param sourceToken: Type of sourceToken. eg. Type<FLOW>
   /// @param targetToken: Type of targetToken. eg. Type<USDC>

>    pub event Swap(

Reviewer 3: Do you have a way to enforce that projects emit these properly? We won't be able to have type requirements with stable cadence any more, but we will be able to emit events with interfaces, so this could be something we can emit directly from this interface. We can actually do that right now

Author: I don't think we can enforce this until the stable cadence, That is why the event is not a part of the interface in this FLIP. I just tried to motivate people to use the unified event wherever is possible.

Reviewer 3: We actually can emit events from interfaces by calling a function that emits the event. Something to consider

        senderAddress: Address,
        receiverAddress: Address?,
        sourceTokenAmount: UFix64,
        receivedTargetTokenAmount: UFix64,
        sourceToken: Type,
        targetToken: Type
    )

```

> 

Reviewer: Two events that would be helpful for dapps and indexing services: Mint and Burn; Mint when new token pair pools are created and Burn when they are destroyed.

Author: Thanks for your valuable feedback. This FLIP intends to focus on APIs to trade however what you said would definitely be beneficial for indexers but those events are highly related to pools and we are not trying to standardise the pool structure and their management in the FLIP. I don't know maybe it would be totally up to the developers how they wanna implement the pools or maybe we can provide an example implementation and unofficially make some recommendations about these things.

	   sourceToTargetTokenPath: Type[],
	   sourceAmount: UFix64
    ): UFix64?

    /// @notice Provides the quotation of the source token amount if user wants to
    /// buy provided targetAmount, i.e. amount of target token.
    ///
    /// If the source to target token path doesn't exists then below function
    /// would return `nil`.
    /// Below function would return the quoted amount after deduction of the fees.
    ///
    /// If the sourceToTargetTokenPath is [Type<FLOW>, Type<BLOCTO>]. 
    /// Where sourceToTargetTokenPath[0] is the source token while 
    /// sourceToTargetTokenPath[sourceToTargetTokenPath.length -1] is 
    /// target token. i.e. FLOW and BLOCTO respectively.
    ///
    /// @param sourceToTargetTokenPath: Offchain computed optimal path from
    ///                                 source token to target token.
    /// @param targetAmount: Amount of target token user wants to buy.
    /// @return Amount of source token user has to pay to buy provided `targetAmount` of target token.
    ///
    pub fun getExactBuyQuoteUsingPath(
        sourceToTargetTokenPath: Type[],
        targetAmount: UFix64
    ): UFix64?
```

### Drawbacks

> There is no real drawbacks because of this FLIP.

Reviewer 3: ```suggestion
There are no real drawbacks because of this FLIP.
```
This isn't necessarily true. One drawback by suggesting this FLIP is that the defi ecosystem will be less flexible to emerge and build its tools organically according to user demand and technical need and we might in the future be stuck with a standard that didn't address some important aspects of defi. 

Granted that is a drawback of most standard proposals, but still maybe worth mentioning because this proposal does kind of preempt any significant demand for defi on Flow from consumers

### Performance Implications

The implementation of the current FLIP would not have any performance implications on the FLOW network or dApps. However, it would decrease the integration time between Automated Market Makers (AMMs) and wallets, thereby facilitating the creation of seamless and composable DeFi dApps.

### Dependencies

There is no external dependencies on this FLIP.

### Engineering Impact

The implementation of the proposed FLIP would have no impact on the FLOW network itself. However, existing AMMs and wallets that have already integrated with AMMs would need to make changes to support the proposed FLIP. Fortunately, since many AMMs are already utilizing a similar interface, it is anticipated that only minimal changes would be required, although these specific modifications fall outside the scope of the current FLIP.

### Best Practices

No changes in the best practices

### Tutorials and Examples

TODO

### Compatibility

TODO

### User Impact

> AMMs on FLOW are required to conform to the proposed FLIP in order to offer a seamless experience to users. Nevertheless, end-users will not perceive any significant changes in the user interface. However, this implementation will simplify the lives of dApp developers and smart contract developers on the FLOW blockchain.

Reviewer 3: ```suggestion
AMMs on FLOW are strongly encouraged to conform to the proposed FLIP in order to offer a seamless experience to users. Nevertheless, end-users will not perceive any significant changes in the user interface. However, this implementation will simplify the lives of dApp developers and smart contract developers on the FLOW blockchain.
```
